infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/NamespaceProvisioner.java

```diff
@@ -0,0 +1,176 @@
+
+import com.google.common.annotations.VisibleForTesting;
+import io.fabric8.kubernetes.api.model.Secret;
+import io.fabric8.kubernetes.api.model.SecretBuilder;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import javax.inject.Inject;
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.model.user.User;
+import org.eclipse.che.api.core.notification.EventSubscriber;
+import org.eclipse.che.api.user.server.PreferenceManager;
+import org.eclipse.che.api.user.server.UserManager;
+import org.eclipse.che.api.user.server.event.PostUserPersistedEvent;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.NamespaceResolutionContext;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesClientFactory;
+import org.eclipse.che.workspace.infrastructure.kubernetes.api.shared.KubernetesNamespaceMeta;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespaceFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class NamespaceProvisioner implements EventSubscriber<PostUserPersistedEvent> {
+  private static final Logger LOG = LoggerFactory.getLogger(NamespaceProvisioner.class);
+  private static final String USER_PROFILE_SECRET_NAME = "user-profile";
+  private static final String USER_PREFERENCES_SECRET_NAME = "user-preferences";
+  private static final String USER_PROFILE_SECRET_MOUNT_PATH = "/config/user/profile";
+  private static final String USER_PREFERENCES_SECRET_MOUNT_PATH = "/config/user/preferences";
+
+  private final KubernetesNamespaceFactory namespaceFactory;
+  private final KubernetesClientFactory clientFactory;
+  private final UserManager userManager;
+  private final PreferenceManager preferenceManager;
+
+  @Inject
+  public NamespaceProvisioner(
+      KubernetesNamespaceFactory namespaceFactory,
+      KubernetesClientFactory clientFactory,
+      UserManager userManager,
+      PreferenceManager preferenceManager) {
+    this.namespaceFactory = namespaceFactory;
+    this.clientFactory = clientFactory;
+    this.userManager = userManager;
+    this.preferenceManager = preferenceManager;
+  }
+
+  public KubernetesNamespaceMeta provision(NamespaceResolutionContext namespaceResolutionContext)
+      throws InfrastructureException {
+
+    KubernetesNamespaceMeta kubernetesNamespaceMeta =
+        namespaceFactory.provision(namespaceResolutionContext);
+    try {
+      createOrUpdateSecrets(userManager.getById(namespaceResolutionContext.getUserId()));
+    } catch (NotFoundException | ServerException e) {
+      LOG.error("Could not find current user. Skipping creation of user information secrets.", e);
```

Comment:
I think we can throw InfrastructureException here. wdyt?

Reply:
Agreed.